wiring/src/spark_wiring_logging.cpp

```diff
@@ -156,50 +173,69 @@ LogLevel spark::LogHandler::categoryLevel(const char *category) const {
     return level;
 }
 
-void spark::LogHandler::logMessage(const char *msg, LogLevel level, const char *category, uint32_t time,
-        const SourceInfo &info) {
-    // Timestamp
+void spark::LogHandler::logMessage(const char *msg, LogLevel level, const char *category, const LogAttributes &attr) {
     char buf[16];
-    snprintf(buf, sizeof(buf), "%010u ", (unsigned)time);
-    write(buf);
-    // Category (optional)
+    // Timestamp
+    if (attr.has_time) {
+        snprintf(buf, sizeof(buf), "%010u ", (unsigned)attr.time);
+        write(buf);
+    }
+    // Category
     if (category) {
         write("[");
         write(category);
         write("] ");
     }
-    // Source file (optional)
-    if (info.file) {
-        write(info.file); // File name
-        write(":");
-        snprintf(buf, sizeof(buf), "%d", info.line); // Line number
-        write(buf);
-        if (info.function) {
+    // Source file
+    if (attr.has_file) {
+        // Strip directory path
+        const char *s = extractFileName(attr.file);
```

Comment:
I was thinking along the same lines initially, but decided to simplify the implementation to avoid unnecessary dynamic allocations, such as in the example below:

```cpp
LogHandler handler(new DefaultLogFormatter(), &Serial1, ...)
```

C++ mixins are also not an option, as they cannot be configured and instantiated dynamically, which is essential for the yet-to-be-developed dynamically enabled logging.

In principle, we can move the formatting code to the existent `StreamLogHandler` to split the functionality more evenly.